src/convert/impls.rs

```diff
@@ -99,14 +99,59 @@ macro_rules! type_wasm_native {
 }
 
 type_wasm_native!(
+    i64 as i64
+    u64 as u64
+    f64 as f64
+);
+
+const F64_SENTINEL: f64 = 4294967297_f64; // 2^32 + 1
```

Comment:
This sentinel is also used for `f32`. `f32` can represent 2^32 exactly but not 2^32+1.

Reply:
I added a comment explaining this.

Reply:
Ah, I see now what you did.
What I meant to propose here is to use `0x1_0000_0000` instead of `0x1_0000_0001`.
Both should not be re-presentable by any 32-bit type, no matter what the actual value actually translates to.

Reply:
> Both should not be re-presentable by any 32-bit type, no matter what the actual value actually translates to.

I'm not sure what you mean here. `f32` can represent 2^32 exactly. [Show case](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=52a59cd0429b2ee58f6c3b67599335ad). 

Please do not make the same mistake I made and think that `f32` can't represent 2^32 exactly just because `println!("{}", 4294967296_f32)` prints `4294967300`. That's the binary floating-point to decimal algorithm making the (valid) assumption that `f32` has ~8 significant decimal digits.

Reply:
Or in other words, 2^32 can't be the sentinel, because:

```js
> Math.fround(2**32) === 2**32
true
```

Reply:
```rust
const F64_SENTINEL: f64 = f64::from_bits(0x1_0000_0000);
```
I mean this, not the integer.

But now that I think about it, I don't actually know how to compare against it in JS.

Reply:
I guess its possible with an `ArrayBuffer`, setting `0x100000000` via `BigUint64Array` and reading it via `Float64Array`, completely unnecessary, let's just go your way.

Reply:
Huh? Why would we use that as the sentinel? I mean, this value *could* be used as the sentinel (because any denormalized float64 value cannot be represented by 32-bit ints or `f32`), but I don't see why we should use this value. Here's the value btw:

```rs
> println!("{}", f64::from_bits(0x1_0000_0000));
0.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002121995791
```

Or put another way, there are give or take 2^62 possible sentinel values we can choose from (namely all `f64` values that cannot be represented by `i32`, `u32`, and `f32`). Why should we use this one?

In my case, I chose 2^32+1, because I wanted a small integer. I also wanted to avoid denomalized values, because they tend to invite floating-point weirdness.

I also want to point out that the bit pattern of the sentinel value doesn't really matter here. It really is about the mathematical number the sentinel represents that we have to get right. So this is a bit different from the classic bit-packing optimizations we are used to from integers.

Reply:
Sorry, I didn't get to see your last comment before making mine.

Reply:
I was explicitly trying pick the sentinel value based on its bit pattern, because it seems more straightforward to me then picking an unrepresentable number by `f32`, which is something you can't tell immediately by looking at it. The bit-pattern is immediately recognizable as not representable.

But this was assuming that we never have to deal with the actual number, like you printed it above.
Unfortunately I couldn't find a way to do this in JS apart from https://github.com/rustwasm/wasm-bindgen/pull/4183#discussion_r1802032668.

Reply:
> The bit-pattern is immediately recognizable as not representable.

Is it? I had to think for a few seconds until I realized that `f64::from_bits(0x1_0000_0000)` is a denorm. I mean, the upper 32 bits containing a 1 bit means pretty much nothing for `f64` in terms of sentinel values. E.g. this is the bit pattern of 42_f64: `0x4045_0000_0000_0000`.

So could you please explain why it is immediately recognizable?

Reply:
> I mean, the upper 32 bits containing a 1 bit means pretty much nothing for `f64` in terms of sentinel values.

Apologies, I see now where the misunderstanding is coming from.

The only reason this is the case is because we convert the `f32` to an `f64` before we send it to JS, instead of sending the actual bit pattern. So this conversion has to make sure that the sentinel value can never be produced.

I was going with the assumption here that we switch that strategy to work with the bit patterns directly and do the necessary conversions on the JS side. What I wasn't aware of, is that JS doesn't make this easy at all, like in Rust. See https://github.com/rustwasm/wasm-bindgen/pull/4183#discussion_r1802032668.

Just for fun and giggles, I will make a small patch where I'm gonna do what I meant.

But yes, this all doesn't make sense when the conversion and comparison in JS has much more overhead.

Reply:
https://github.com/rustwasm/wasm-bindgen/commit/fbd21523e6089a50a509ed72d71fc2c44453e8fb

While this could be cleaned up, I think its clear that using bitpacking adds no real value while introducing comparatively tons of overhead.

Reply:
Ah I see. In that case, using u64 for the ABI probably would make your approach easier to implement, since you'd be able to use regular bit shifts and so on just like in Rust. The only tricky part would be to get the bits of a float32, but you already devised a solution for that.

The reason why I did not want u64 as the ABI is that u64 requires bigint on the JS side. JS bigints are notoriously slow compared to JS numbers (sometimes 10x slower), and this was supposed to be a performance optimization.

Reply:
Indeed, that's why I was trying to stick with `f64`.
But yeah, I think we can conclude this experiment!